src/storage/src/hummock/store/version.rs

```diff
@@ -183,29 +189,196 @@ impl StagingVersion {
+        Self {
+            table_id,
+            scope: max_committed_epoch + 1,
+            levels: vec![],
+        }
+    }
+
+    fn scope(&self) -> HummockEpoch {
+        self.scope
+    }
+
+    fn get_indices(
+        &self,
+        table_view_position: usize,
+        sub_level_id: HummockEpoch,
+    ) -> Option<&Vec<usize>> {
+        if sub_level_id >= self.scope {
+            Some(&self.levels[table_view_position].1 .0)
+        } else {
+            None
+        }
+    }
+
+    fn construct_index(
+        &self,
+        sub_level: &Level,
+        object_ids: impl Iterator<Item = HummockSstableObjectId>,
+    ) -> PrunedVersionLevel {
+        let mut hexff_indices = object_ids
+            .map(|object_id| (usize::MAX, object_id))
+            .collect_vec();
+        hexff_indices.sort_by_key(|(_usize_max, object_id)| *object_id);
+        for (index, table_info) in sub_level.get_table_infos().iter().enumerate() {
+            if let Ok(position) = hexff_indices.binary_search_by_key(&table_info.object_id, |(_, object_id)| *object_id) && table_info.get_table_ids().binary_search(&self.table_id.table_id).is_ok() {
+                hexff_indices[position].0 = index;
+            }
+        }
+        hexff_indices.retain(|(index, _)| *index != usize::MAX);
+        hexff_indices.sort();
+        hexff_indices.into_iter().unzip()
+    }
+
+    fn apply(
+        &self,
+        committed_version: &CommittedVersion,
+        mut cleaned_staging_ssts: Vec<(HummockEpoch, Vec<HummockSstableObjectId>)>,
+    ) -> Self {
+        cleaned_staging_ssts.sort_by_key(|(epoch, _)| *epoch);
+
+        let committed_l0_sub_levels = committed_version.l0_sublevels(self.table_id);
+        let mut sub_level_asc_iter = committed_l0_sub_levels.iter().rev().peekable();
+        let mut new_levels = vec![];
+        for (old_sub_level_id, (_, old_object_ids)) in self
+            .levels
+            .iter()
+            .rev()
+            .filter(|(_, (indices, _))| !indices.is_empty())
+        {
+            while let Some(sub_level) = sub_level_asc_iter.peek() && sub_level.sub_level_id < *old_sub_level_id {
+                sub_level_asc_iter.next();
+            }
+            match sub_level_asc_iter.peek() {
+                Some(sub_level) => {
+                    if sub_level.sub_level_id == *old_sub_level_id
+                        && sub_level.level_type() != LevelType::Nonoverlapping
+                    {
+                        let new_indices =
+                            self.construct_index(sub_level, old_object_ids.iter().copied());
+                        new_levels.push((*old_sub_level_id, new_indices));
+
+                        sub_level_asc_iter.next();
+                    }
+                }
+                None => break,
+            }
+        }
+
+        let mut levels_to_extend = vec![];
+        for (idx, sub_level) in committed_l0_sub_levels.iter().enumerate() {
```

Comment:
Could we break this loop before check `cleaned_staging_ssts`? For example:
```rust
 for (idx, sub_level) in committed_l0_sub_levels.iter().enumerate() {
     if sub_level.level_type != LevelType::Overlapping {
         break;
      }

```

Reply:
> Could we break this loop before check `cleaned_staging_ssts`? For example:
> 
> ```rust
>  for (idx, sub_level) in committed_l0_sub_levels.iter().enumerate() {
>      if sub_level.level_type != LevelType::Overlapping {
>          break;
>       }
> ```

No. Because overlapping levels may also exist under nonoverlapping levels.